File: packages/graph-grammar/src/rewrite.ts
import type {
  Rule,
  Match,
  PropExpr,
  PropValue,
  Props,
  GEdge,
  EmbeddingRule,
} from './types.ts'
import { GraphIndex, LabelBucket, makeNode, makeEdge } from './graph.ts'
import { RNG } from './util.ts'

// ============================================================================
// Rewriting (the "double pushout"-flavoured rewrite step, simplified).
//
// Given a host GraphIndex, a Rule, and a Match, produce the rewritten graph
// in place on the index. We follow these phases:
//   1. Resolve which LHS nodes are *preserved* (referenced by a RHS node via
//      mapFrom) versus *deleted*.
//   2. Compute RHS-node -> host-node resolution (preserved → existing host
//      node, created → fresh node).
//   3. Collect dangling edges of deleted LHS-matched nodes and apply embedding.
//   4. Delete matched edges/nodes that are not preserved.
//   5. Relabel / apply property mutations to preserved nodes.
//   6. Create new RHS nodes and edges.
// ============================================================================

export interface RewriteContext {
  rng: RNG;
  /** global counter for the `counter` PropExpr. */
  counter: { value: number };
}

function evalExpr (
  expr: PropExpr,
  ctx: RewriteContext,
  host: GraphIndex,
  nodeMap: Record<string, string>
): PropValue {
  switch (expr.kind) {
    case 'literal':
      return expr.value
    case 'copy': {
      const hid = nodeMap[expr.from]
      const n = hid ? host.nodes.get(hid) : undefined
      return n ? (n.props[expr.key] ?? null) : null
    }
    case 'increment': {
      const hid = nodeMap[expr.from]
      const n = hid ? host.nodes.get(hid) : undefined
      const base = n && typeof n.props[expr.key] === 'number' ? (n.props[expr.key] as number) : 0
      return base + expr.by
    }
    case 'randInt':
      return ctx.rng.int(expr.min, expr.max)
    case 'randFloat':
      return ctx.rng.float(expr.min, expr.max)
    case 'counter':
      return ++ctx.counter.value
    default:
      return null
  }
}

function applySetProps (
  base: Props,
  setProps: Record<string, PropExpr> | undefined,
  ctx: RewriteContext,
  host: GraphIndex,
  nodeMap: Record<string, string>
): Props {
  if (!setProps) return base
  const out: Props = { ...base }
  for (const [k, expr] of Object.entries(setProps)) {
    out[k] = evalExpr(expr, ctx, host, nodeMap)
  }
  return out
}

export interface ApplyResult {
  createdNodes: string[];
  createdEdges: string[];
  deletedNodes: string[];
  deletedEdges: string[];
}

/**
 * Apply `rule` at `match` to `host` (mutating the index). Returns the set of
 * changed element ids for highlighting/animation.
 */
export function applyRule (
  host: GraphIndex,
  rule: Rule,
  match: Match,
  ctx: RewriteContext
): ApplyResult {
  const nodeMap = match.nodeMap // LHS node id -> host node id
  const created: string[] = []
  const createdEdges: string[] = []
  const deletedNodes: string[] = []
  const deletedEdges: string[] = []

  // 1. Which LHS nodes are preserved?
  const preservedLhs = new Set<string>()
  for (const rn of rule.rhs.nodes) if (rn.mapFrom) preservedLhs.add(rn.mapFrom)

  const matchedHostNodes = new Set(Object.values(nodeMap))

  // 2. RHS node -> host node resolution. Build it before edge work.
  const rhsToHost = new Map<string, string>()

  // Preserved nodes first (relabel + props).
  for (const rn of rule.rhs.nodes) {
    if (rn.mapFrom && nodeMap[rn.mapFrom]) {
      const hid = nodeMap[rn.mapFrom]
      const hn = host.nodes.get(hid)
      if (hn) {
        // relabel if RHS gives a concrete (non-empty) label different from host
        if (rn.label && rn.label !== '*' && rn.label !== hn.label) {
          // label change requires reindexing in byLabel
          host.byLabel.get(hn.label)?.delete(hid)
          hn.label = rn.label
          let bucket = host.byLabel.get(hn.label)
          if (!bucket) host.byLabel.set(hn.label, (bucket = new LabelBucket()))
          bucket.add(hid)
        }
        hn.props = applySetProps(hn.props, rn.setProps, ctx, host, nodeMap)
        rhsToHost.set(rn.id, hid)
      }
    }
  }

  // 3. Dangling edges of soon-to-be-deleted nodes → embedding.
  const lhsDeleted: string[] = []
  for (const [lhsId, hid] of Object.entries(nodeMap)) {
    if (!preservedLhs.has(lhsId)) lhsDeleted.push(hid)
  }

  // matched host edge ids (so we don't treat them as "external" dangling edges)
  const matchedHostEdgeIds = new Set(Object.values(match.edgeMap))

  // Collect dangling edges grouped by which LHS node caused them.
  interface Dangling {
    edge: GEdge;
    lhsNodeId: string;
    externalNodeId: string;
    danglingIsSource: boolean; // is the deleted node the source side?
  }
  const danglings: Dangling[] = []
  const lhsByHost = new Map<string, string>()
  for (const [lhsId, hid] of Object.entries(nodeMap)) lhsByHost.set(hid, lhsId)

  for (const lhsId of Object.keys(nodeMap)) {
    if (preservedLhs.has(lhsId)) continue
    const hid = nodeMap[lhsId]
    for (const e of host.incidentEdges(hid)) {
      if (matchedHostEdgeIds.has(e.id)) continue // internal to match, handled by delete
      const otherId = e.source === hid ? e.target : e.source
      // If the other endpoint is also a deleted matched node, this edge simply dies.
      if (matchedHostNodes.has(otherId) && !preservedLhs.has(lhsByHost.get(otherId) ?? '')) {
        continue
      }
      danglings.push({
        edge: e,
        lhsNodeId: lhsId,
        externalNodeId: otherId,
        danglingIsSource: e.source === hid,
      })
    }
  }

  // 6a. Create new RHS nodes (those without a mapFrom). Seed their layout near
  //     the centroid of the match so a renderer has a sensible starting point
  //     (positions are optional layout metadata; physics is the renderer's job).
  let cx = 0
  let cy = 0
  let cn = 0
  for (const hid of matchedHostNodes) {
    const hn = host.nodes.get(hid)
    if (hn && hn.x != null && hn.y != null) {
      cx += hn.x
      cy += hn.y
      cn++
    }
  }
  if (cn) {
    cx /= cn
    cy /= cn
  }

  for (const rn of rule.rhs.nodes) {
    if (rn.mapFrom && nodeMap[rn.mapFrom]) continue // preserved, already done
    const props = applySetProps(rn.props ?? {}, rn.setProps, ctx, host, nodeMap)
    const jitter = () => (ctx.rng.next() - 0.5) * 40
    const node = makeNode(
      rn.label && rn.label !== '*' ? rn.label : 'node',
      props,
      cn ? cx + jitter() : (rn.x ?? jitter()),
      cn ? cy + jitter() : (rn.y ?? jitter())
    )
    host.addNode(node)
    rhsToHost.set(rn.id, node.id)
    created.push(node.id)
  }

  // 4. Delete matched edges that are not preserved by RHS.
  const preservedHostEdges = new Set<string>()
  for (const re of rule.rhs.edges) {
    if (re.mapFrom && match.edgeMap[re.mapFrom]) {
      preservedHostEdges.add(match.edgeMap[re.mapFrom])
    }
  }
  for (const hostEdgeId of Object.values(match.edgeMap)) {
    if (!preservedHostEdges.has(hostEdgeId)) {
      if (host.edges.has(hostEdgeId)) {
        host.removeEdge(hostEdgeId)
        deletedEdges.push(hostEdgeId)
      }
    }
  }

  // 5/3. Apply embedding for dangling edges, then delete the LHS-deleted nodes.
  const embByLhs = new Map<string, EmbeddingRule>()
  for (const er of rule.embedding) embByLhs.set(er.lhsNodeId, er)

  for (const d of danglings) {
    const er = embByLhs.get(d.lhsNodeId)
    const strategy = er?.strategy ?? 'remove'
    if (er?.edgeLabelFilter && d.edge.label !== er.edgeLabelFilter) {
      // not governed: default remove
      continue
    }
    if (strategy === 'remove') continue
    const newLabel = er?.newEdgeLabel || d.edge.label
    const targets: string[] = []
    if (strategy === 'redirectTo' && er?.targetRhsNodeId) {
      const h = rhsToHost.get(er.targetRhsNodeId)
      if (h) targets.push(h)
    } else if (strategy === 'redirectToAll') {
      for (const id of created) targets.push(id)
    }
    for (const t of targets) {
      const src = d.danglingIsSource ? t : d.externalNodeId
      const tgt = d.danglingIsSource ? d.externalNodeId : t
      const ne = makeEdge(src, tgt, newLabel, d.edge.directed, { ...d.edge.props })
      host.addEdge(ne)
      createdEdges.push(ne.id)
    }
  }

  // Now delete the LHS-deleted host nodes (removes remaining incident edges).
  for (const hid of lhsDeleted) {
    if (host.nodes.has(hid)) {
      for (const e of host.incidentEdges(hid)) {
        if (host.edges.has(e.id)) deletedEdges.push(e.id)
      }
      host.removeNode(hid)
      deletedNodes.push(hid)
    }
  }

  // 6b. Create new RHS edges (not preserved). Resolve endpoints via rhsToHost.
  for (const re of rule.rhs.edges) {
    if (re.mapFrom && preservedHostEdges.has(match.edgeMap[re.mapFrom] ?? '')) {
      // preserved edge: optionally relabel and merge any setProps onto it.
      const hostId = match.edgeMap[re.mapFrom]
      const he = host.edges.get(hostId)
      if (he) {
        if (re.label && re.label !== '*' && re.label !== he.label) he.label = re.label
        he.props = applySetProps(he.props, re.setProps, ctx, host, nodeMap)
      }
      continue
    }
    const s = rhsToHost.get(re.source)
    const t = rhsToHost.get(re.target)
    if (!s || !t) continue // endpoint not resolvable (shouldn't happen)
    const props = applySetProps({ ...re.props }, re.setProps, ctx, host, nodeMap)
    const ne = makeEdge(s, t, re.label === '*' ? '' : re.label, re.directed, props)
    host.addEdge(ne)
    createdEdges.push(ne.id)
  }

  return { createdNodes: created, createdEdges, deletedNodes, deletedEdges }
}

File: packages/graph-grammar-native/src/rewrite.rs
use std::collections::{HashMap, HashSet};

use indexmap::IndexMap;
use serde_json::Value;

use crate::graph::Host;
use crate::rng::{number_value, Rng};
use crate::types::{ApiError, EmbeddingRule, GEdge, GNode, MatchResult, PropExpr, Props, Rule};

// ============================================================================
// Property expressions (rewrite.ts evalExpr) ,deterministic kinds only.
// ============================================================================

/// The no-seed (deterministic) entry point has no RNG, so a random PropExpr is a
/// usage error there (code 3); the seeded entry point always supplies one.
fn need_rng(rng: Option<&mut Rng>) -> Result<&mut Rng, ApiError> {
    rng.ok_or(ApiError {
        code: 3,
        detail: "random PropExpr (randInt/randFloat) requires a seed; use gg_apply_rule_seeded"
            .to_string(),
    })
}

fn eval_expr(
    expr: &PropExpr,
    host: &Host,
    node_map: &IndexMap<String, String>,
    rng: Option<&mut Rng>,
    counter: &mut i64,
) -> Result<Value, ApiError> {
    let lookup = |from: &str, key: &str| -> Option<Value> {
        let hid = node_map.get(from)?;
        let n = host.nodes.get(hid)?;
        Some(n.props.get(key).cloned().unwrap_or(Value::Null))
    };
    Ok(match expr {
        PropExpr::Literal { value } => value.clone(),
        PropExpr::Copy { from, key } => lookup(from, key).unwrap_or(Value::Null),
        PropExpr::Increment { from, key, by } => {
            let base = lookup(from, key).and_then(|v| v.as_f64()).unwrap_or(0.0);
            number_value(base + by)
        }
        PropExpr::Counter => {
            *counter += 1;
            Value::from(*counter)
        }
        PropExpr::RandInt { min, max } => number_value(need_rng(rng)?.int(*min, *max)),
        PropExpr::RandFloat { min, max } => number_value(need_rng(rng)?.float(*min, *max)),
    })
}

fn apply_set_props(
    base: &Props,
    set_props: &Option<IndexMap<String, PropExpr>>,
    host: &Host,
    node_map: &IndexMap<String, String>,
    mut rng: Option<&mut Rng>,
    counter: &mut i64,
) -> Result<Props, ApiError> {
    let mut out = base.clone();
    if let Some(sp) = set_props {
        // Insertion order (mirrors the engine's Object.entries), so the RNG is
        // consumed in the same order for multiple random expressions.
        for (k, expr) in sp {
            out.insert(k.clone(), eval_expr(expr, host, node_map, rng.as_deref_mut(), counter)?);
        }
    }
    Ok(out)
}

// ============================================================================
// Rewrite ,a faithful port of rewrite.ts applyRule, minus the RNG layout
// jitter for created nodes (positions are renderer metadata). Created element
// ids are deterministic (`n_native_*` / `e_native_*`); the TS engine uses
// volatile uid()s, so conformance compares created elements up to id-renaming.
// ============================================================================

pub(crate) struct ApplyOut {
    pub(crate) created_nodes: Vec<String>,
    pub(crate) created_edges: Vec<String>,
    pub(crate) deleted_nodes: Vec<String>,
    pub(crate) deleted_edges: Vec<String>,
}

pub(crate) fn apply_rule_at(
    host: &mut Host,
    rule: &Rule,
    m: &MatchResult,
    mut rng: Option<&mut Rng>,
    counter: &mut i64,
    id_seq: &mut i64,
) -> Result<ApplyOut, ApiError> {
    // id_seq is owned by the caller so created-element ids stay unique ACROSS
    // applies (the engine's uid() is globally monotonic); a per-call counter
    // would collide on every engine step and corrupt the graph.
    let mut next_id = |prefix: &str| {
        *id_seq += 1;
        format!("{prefix}_native_{}", *id_seq)
    };

    let mut created_nodes: Vec<String> = Vec::new();
    let mut created_edges: Vec<String> = Vec::new();
    let mut deleted_nodes: Vec<String> = Vec::new();
    let mut deleted_edges: Vec<String> = Vec::new();

    // 1. Which LHS nodes are preserved (referenced by a RHS node via mapFrom)?
    let preserved_lhs: HashSet<String> =
        rule.rhs.nodes.iter().filter_map(|rn| rn.map_from.clone()).collect();
    let matched_host_nodes: HashSet<String> = m.node_map.values().cloned().collect();
    let mut rhs_to_host: HashMap<String, String> = HashMap::new();

    // 2. Preserved nodes: relabel + property mutations.
    for rn in &rule.rhs.nodes {
        let Some(mf) = &rn.map_from else { continue };
        let Some(hid) = m.node_map.get(mf) else { continue };
        if !host.nodes.contains_key(hid) {
            continue;
        }
        if !rn.label.is_empty() && rn.label != "*" {
            host.relabel_node(hid, &rn.label);
        }
        let base = host.nodes[hid].props.clone();
        let new_props = apply_set_props(&base, &rn.set_props, host, &m.node_map, rng.as_deref_mut(), counter)?;
        host.nodes.get_mut(hid).unwrap().props = new_props;
        rhs_to_host.insert(rn.id.clone(), hid.clone());
    }

    // 3. Collect dangling edges of soon-to-be-deleted matched nodes.
    let lhs_by_host: HashMap<String, String> =
        m.node_map.iter().map(|(l, h)| (h.clone(), l.clone())).collect();
    let matched_host_edge_ids: HashSet<String> = m.edge_map.values().cloned().collect();

    struct Dangling {
        edge: GEdge,
        lhs_node_id: String,
        external: String,
        is_source: bool,
    }
    let mut danglings: Vec<Dangling> = Vec::new();
    for (lid, hid) in &m.node_map {
        if preserved_lhs.contains(lid) {
            continue;
        }
        for e in host.incident_edges(hid) {
            if matched_host_edge_ids.contains(&e.id) {
                continue;
            }
            let other = if &e.source == hid { e.target.clone() } else { e.source.clone() };
            let other_is_deleted_match = matched_host_nodes.contains(&other)
                && !preserved_lhs.contains(lhs_by_host.get(&other).map(|s| s.as_str()).unwrap_or(""));
            if other_is_deleted_match {
                continue;
            }
            let is_source = &e.source == hid;
            danglings.push(Dangling { edge: e, lhs_node_id: lid.clone(), external: other, is_source });
        }
    }

    // Centroid of matched nodes that carry layout (mirrors rewrite.ts 6a) ,seeds
    // created-node positions. Only consumed by jitter on the stochastic path.
    let (mut cx, mut cy, mut cn) = (0.0_f64, 0.0_f64, 0u32);
    for hid in &matched_host_nodes {
        if let Some(hn) = host.nodes.get(hid) {
            if let (Some(x), Some(y)) = (hn.x, hn.y) {
                cx += x;
                cy += y;
                cn += 1;
            }
        }
    }
    if cn > 0 {
        cx /= cn as f64;
        cy /= cn as f64;
    }

    // 4. Create new RHS nodes (those without a mapFrom).
    for rn in &rule.rhs.nodes {
        if let Some(mf) = &rn.map_from {
            if m.node_map.contains_key(mf) {
                continue;
            }
        }
        let props = apply_set_props(&rn.props, &rn.set_props, host, &m.node_map, rng.as_deref_mut(), counter)?;
        // Layout jitter: `(next() - 0.5) * 40`, x then y, drawing only when there
        // is no provided coordinate ,exactly as `cn ? c+jitter() : (rn.x ?? jitter())`.
        let (x, y) = match rng.as_deref_mut() {
            Some(r) => {
                let x = if cn > 0 {
                    cx + (r.next() - 0.5) * 40.0
                } else {
                    rn.x.unwrap_or_else(|| (r.next() - 0.5) * 40.0)
                };
                let y = if cn > 0 {
                    cy + (r.next() - 0.5) * 40.0
                } else {
                    rn.y.unwrap_or_else(|| (r.next() - 0.5) * 40.0)
                };
                (Some(x), Some(y))
            }
            None => (None, None),
        };
        let id = next_id("n");
        let label = if !rn.label.is_empty() && rn.label != "*" {
            rn.label.clone()
        } else {
            "node".to_string()
        };
        host.add_node(GNode { id: id.clone(), label, props, x, y });
        rhs_to_host.insert(rn.id.clone(), id.clone());
        created_nodes.push(id);
    }

    // 5. Delete matched host edges not preserved by the RHS.
    let preserved_host_edges: HashSet<String> = rule
        .rhs
        .edges
        .iter()
        .filter_map(|re| re.map_from.as_ref().and_then(|mf| m.edge_map.get(mf)).cloned())
        .collect();
    for he_id in m.edge_map.values() {
        if !preserved_host_edges.contains(he_id) && host.edges.contains_key(he_id) {
            host.remove_edge(he_id);
            deleted_edges.push(he_id.clone());
        }
    }

    // 6. Embedding for dangling edges.
    let emb_by_lhs: HashMap<&str, &EmbeddingRule> =
        rule.embedding.iter().map(|er| (er.lhs_node_id.as_str(), er)).collect();
    for d in &danglings {
        let er = emb_by_lhs.get(d.lhs_node_id.as_str()).copied();
        if let Some(e) = er {
            if let Some(f) = &e.edge_label_filter {
                if &d.edge.label != f {
                    continue;
                }
            }
        }
        let strategy = er.map(|e| e.strategy.as_str()).unwrap_or("remove");
        if strategy == "remove" {
            continue;
        }
        let new_label = er
            .and_then(|e| e.new_edge_label.clone())
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| d.edge.label.clone());
        let mut targets: Vec<String> = Vec::new();
        if strategy == "redirectTo" {
            if let Some(tr) = er.and_then(|e| e.target_rhs_node_id.as_ref()) {
                if let Some(h) = rhs_to_host.get(tr) {
                    targets.push(h.clone());
                }
            }
        } else if strategy == "redirectToAll" {
            targets.extend(created_nodes.iter().cloned());
        }
        for t in targets {
            let (src, tgt) = if d.is_source {
                (t.clone(), d.external.clone())
            } else {
                (d.external.clone(), t.clone())
            };
            let id = next_id("e");
            host.add_edge(GEdge {
                id: id.clone(),
                source: src,
                target: tgt,
                label: new_label.clone(),
                props: d.edge.props.clone(),
                directed: d.edge.directed,
            });
            created_edges.push(id);
        }
    }

    // 7. Delete the LHS-deleted host nodes (removes remaining incident edges).
    let lhs_deleted: Vec<String> = m
        .node_map
        .iter()
        .filter(|(lid, _)| !preserved_lhs.contains(*lid))
        .map(|(_, hid)| hid.clone())
        .collect();
    for hid in &lhs_deleted {
        if host.nodes.contains_key(hid) {
            for e in host.incident_edges(hid) {
                if host.edges.contains_key(&e.id) {
                    deleted_edges.push(e.id.clone());
                }
            }
            host.remove_node(hid);
            deleted_nodes.push(hid.clone());
        }
    }

    // 8. Create new RHS edges (and relabel/merge preserved ones).
    for re in &rule.rhs.edges {
        if let Some(mf) = &re.map_from {
            if let Some(hid) = m.edge_map.get(mf) {
                if preserved_host_edges.contains(hid) {
                    if !re.label.is_empty() && re.label != "*" {
                        if let Some(he) = host.edges.get_mut(hid) {
                            if re.label != he.label {
                                he.label = re.label.clone();
                            }
                        }
                    }
                    if re.set_props.is_some() {
                        let base = host.edges[hid].props.clone();
                        let np = apply_set_props(&base, &re.set_props, host, &m.node_map, rng.as_deref_mut(), counter)?;
                        host.edges.get_mut(hid).unwrap().props = np;
                    }
                    continue;
                }
            }
        }
        let (Some(s), Some(t)) = (rhs_to_host.get(&re.source), rhs_to_host.get(&re.target)) else {
            continue;
        };
        let (s, t) = (s.clone(), t.clone());
        let props = apply_set_props(&re.props, &re.set_props, host, &m.node_map, rng.as_deref_mut(), counter)?;
        let id = next_id("e");
        let label = if re.label == "*" { String::new() } else { re.label.clone() };
        host.add_edge(GEdge { id: id.clone(), source: s, target: t, label, props, directed: re.directed });
        created_edges.push(id);
    }

    Ok(ApplyOut { created_nodes, created_edges, deleted_nodes, deleted_edges })
}
